central/pruning/pruning.go

```diff
@@ -485,6 +494,140 @@ func (g *garbageCollectorImpl) collectImages(config *storage.PrivateConfig) {
+	clusters, err := g.clusters.SearchRawClusters(pruningCtx, query)
+	if err != nil {
+		log.Errorf("[Cluster Pruning] error searching for clusters: %v", err)
+		return
+	}
+
+	log.Infof("[Cluster Pruning] found %d cluster(s) that haven't been active in over %d days", len(clusters), retentionDays)
+
+	clustersToPrune := make([]string, 0)
+	for _, cluster := range clusters {
+		if sliceutils.MapsIntersect(clusterRetention.GetIgnoreClusterLabels(), cluster.GetLabels()) {
+			log.Infof("[Cluster Pruning] skipping excluded cluster with id %s", cluster.GetId())
+			continue
+		}
+
+		// Don't delete if the cluster contains central
+		hasCentral, err := g.checkIfClusterContainsCentral(cluster)
+		if err != nil {
+			log.Errorf("[Cluster Pruning] error searching for deployements in cluster: %v", err)
+			return
+		}
+		if hasCentral {
+			// Warning because it's important to know that your central cluster is unhealthy
+			log.Warnf("[Cluster Pruning] skipping pruning cluster with id %s because this cluster contains the central deployment.", cluster.GetId())
+			continue
+		}
+		clustersToPrune = append(clustersToPrune, cluster.GetId())
+	}
+
+	if allClusterCount == len(clustersToPrune) {
+		log.Warnf("[Cluster Pruning] skipping pruning because all %d cluster(s) are unhealthy and this is an abnormal state. Please remove any manually if desired.", allClusterCount)
+		return
+	}
+
+	if len(clustersToPrune) == 0 {
+		// Debug log as it's be noisy, but is helpful if debugging
+		log.Debug("[Cluster Pruning] no inactive clusters found.")
+		return
+	}
+
+	for _, clusterId := range clustersToPrune {
+		log.Infof("[Cluster Pruning] Removing cluster with ID %s", clusterId)
+		if err := g.clusters.RemoveCluster(pruningCtx, clusterId, nil); err != nil {
+			log.Error(err)
+			return
+		}
+	}
+}
+
+func (g *garbageCollectorImpl) checkIfClusterContainsCentral(cluster *storage.Cluster) (bool, error) {
+	//This query could be expensive, but it's a rare occurrence. It only happens if there is a cluster that has been unhealthy for a long time (configurable)
+	query := search.NewQueryBuilder().
+		AddStrings(search.ClusterID, cluster.GetId()).
+		AddExactMatches(search.DeploymentName, "central")
+	deploys, err := g.deployments.SearchRawDeployments(pruningCtx, query.ProtoQuery())
+	if err != nil {
+		return false, err
+	}
+
+	// As long as we find at least one deployment that matches the criteria
+	// This is meant to be tolerant of false positives while extremely intolerant of false negatives. Rather not delete a cluster than to delete one with central in it.
+	for _, d := range deploys {
+		// While this could be done via the searcher, this was moved out to reduce complexity of the search.
+		// The search might run more often, but this manual grep only happens if the cluster is unhealthy AND has >= 1 deployment with name central.
+		if d.GetLabels()["app"] == "central" && d.GetAnnotations()["owner"] == "stackrox" {
+			return true, nil
```

Comment:
Are deployment `name = "central" && label "app=central" exists &&  annotation "owner=stackrox" exists` necessary conditions for StackRox central deployment?